tools/lockstat.py

```diff
@@ -0,0 +1,403 @@
+#!/usr/bin/python
+#
+# lockstats traces lock events and display locks statistics.
+#
+# USAGE: lockstats
+#
+
+from __future__ import print_function
+from bcc import BPF, USDT
+import argparse
+import subprocess
+import ctypes as ct
+from time import sleep, strftime
+from datetime import datetime, timedelta
+
+examples = """
+    lockstats                           # trace system wide
+    lockstats -d 5                      # trace for 5 seconds only
+    lockstats -i 5                      # display stats every 5 seconds
+    lockstats -p 123                    # trace locks for PID 123
+    lockstats -t 321                    # trace locks for PID 321
+    lockstats -c pipe_                  # display stats only for lock callers with 'pipe_' substring
+    lockstats -S acq_count              # sort lock acquired results on acquired count
+    lockstats -S hld_total              # sort lock held results on total held time
+    lockstats -S acq_count,hld_total    # combination of above
+    lockstats -n 3                      # display 3 locks
+    lockstats -s 3                      # display 3 levels of stack
+"""
+
+parser = argparse.ArgumentParser(
+    description="",
+    formatter_class=argparse.RawDescriptionHelpFormatter,
+    epilog=examples)
+parser.add_argument("-d", "--duration", type=int,
+    help="total duration of trace in seconds")
+parser.add_argument("-i", "--interval", type=int,
+    help="print summary at this interval (seconds)")
+parser.add_argument("-n", "--locks", type=int, default=999999999,
+    help="print given number of locks")
+parser.add_argument("-s", "--stacks", type=int, default=1,
+    help="print given number of stack entries")
+parser.add_argument("-c", "--caller",
+    help="print locks taken by given caller")
+parser.add_argument("-S", "--sort",
+    help="sort data on <aq_field,hd_field>, fields: acq_[max|total|count] hld_[max|total|count]")
+parser.add_argument("-p", "--pid",
+    help="trace this PID only")
+parser.add_argument("-t", "--tid",
+    help="trace this TID only")
+
+args = parser.parse_args()
+
+if args.duration and args.interval:
+    print("Options -d and -i are mutualy exclusive")
+    exit(-1)
+
+program = """
+#include <uapi/linux/ptrace.h>
+
+struct depth_id {
+  u64 id;
+  u64 depth;
+};
+
+BPF_ARRAY(enabled,   u64, 1);
+BPF_HASH(track,      u64, u64);
+BPF_HASH(time,       u64, u64);
+BPF_HASH(lock_depth, u64, u64);
+BPF_HASH(time_held,  struct depth_id, u64);
+BPF_HASH(stack,      struct depth_id, u64);
+
+BPF_HASH(aq_report_count, u64, u64);
+BPF_HASH(aq_report_max,   u64, u64);
+BPF_HASH(aq_report_total,   u64, u64);
+
+BPF_HASH(hl_report_count, u64, u64);
+BPF_HASH(hl_report_max,   u64, u64);
+BPF_HASH(hl_report_total,   u64, u64);
+
+BPF_STACK_TRACE(stack_traces, 10240);
```

Comment:
inside the kernel, the number of entries will be rounded up to power of 2.
https://github.com/torvalds/linux/blob/master/kernel/bpf/stackmap.c#L114
could you just round up to power of 2 here to avoid confusion.